distributed_shampoo/tests/shampoo_types_test.py

```diff
@@ -21,7 +21,7 @@
 class AdaGradGraftingConfigSubclassesTest(unittest.TestCase):
     def test_illegal_epsilon(self) -> None:
         epsilon = 0.0
-        for cls in AdaGradGraftingConfig.__subclasses__():
+        for cls in [AdaGradGraftingConfig] + AdaGradGraftingConfig.__subclasses__():
```

Comment:
Now that I'm realizing how `__subclasses__` works, don't we want to search all subclasses recursively? The current code only covers classes that directly inherit from the base class, but we expect every child class to have the same field and `__post_init__` function. Of course it could in principle be overwritten and then the test has to be manually adjusted, but I don't think we currently have such a case in the code base.

Reply:
> __subclasses__ does search all subclasses recursively.

I'm confused, according to the [docs](https://docs.python.org/3.10/library/stdtypes.html?highlight=__subclasses__#class.__subclasses__) it doesn't. Also, see this example:
```python
class Animal:
    pass

class Mammal(Animal):
    pass

class Dog(Mammal):
    pass

class Cat(Mammal):
    pass

print(Animal.__subclasses__())   # -> [<class '__main__.Mammal'>]
print(Mammal.__subclasses__())   # -> [<class '__main__.Dog'>, <class '__main__.Cat'>]
```

> About the situation if a subclass overwrites the __post_init__() and say removing the assertion, tests here will fail because it did not throw the assertion error. That is exactly what we want because the people who adds the subclass will notice this.

Agreed.